### PR TITLE
Fix #4519: shadow to srw

### LINKED ISSUE #4519: Shadow to SRW conversion
Add a "Open as a new SRW Simulation" feature to Shadow, which reverses the current conversion from SRW to Shadow.

### PR DESCRIPTION
**Pull request by gurhar1133**
SRWShadowConverter now converts to srw
code sharing in UI conversion modal
test added

@moellep you may want to pull the branch and test it out a bit

### CODE REVIEW

**Review by robnagler — CHANGES_REQUESTED** [review]
I can't comment on how it works

**Review by mkeilman — CHANGES_REQUESTED** [review]
mostly style comments, but be sure to delete `srdbg()`

File: sirepo/template/srw_shadow_converter.py [hunk 1]
```diff
@@ -150,14 +149,65 @@ class SRWShadowConverter():
...
+        res = simulation_db.default_data(sirepo.sim_data.get_class('srw').sim_type())
+        self.beamline = res.models.beamline
+        self.__sim_to_srw(data, res.models)
+        res.models.simulation.sourceType = data.simulation.sourceType
+        if res.models.simulation.sourceType == 'u':
+            self.__undulator_to_srw(data, res.models)
+        elif res.models.simulation.sourceType == 'g':
+            self.__geometric_source_to_srw(data, res.models)
+        elif res.models.simulation.sourceType == 'm':
+            self.__multipole_to_srw(data, res.models)
```
[the remark this answers is not in the document]
Comment (gurhar1133) [reply sirepo/template/srw_shadow_converter.py:203]: Nice! I was wondering about this part

**Comment by gurhar1133** [comment]
> It's going to be impossible for me to review this because of the merge with master. 257 files is too much.

haha! fair point. I'll revert to the commit prior to when i ran fmt.

**Comment by gurhar1133** [comment]
@robnagler I reverted the format and just formatted the relevant .py files

File: sirepo/package_data/static/js/sirepo-components.js [hunk 2]
```diff
@@ -3007,6 +3007,85 @@ SIREPO.app.directive('commonFooter', function() {
...
+                if (! res.name){
+                    res.name = 'newSim';
+                }
+                return res;
+            }
+
+            function genSimURL(data) {
+                $scope.newSimURL = requestSender.formatUrlLocal(
+                    'beamline',
+                    {
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo-components.js:3059]: missing trailing comma, but i personally would write it like this: `{ 'simulationId': data.models.simulation.simulationId}`. There only one element.

File: sirepo/package_data/static/js/sirepo-components.js [hunk 3]
```diff
@@ -3007,6 +3007,85 @@ SIREPO.app.directive('commonFooter', function() {
...
+                        'simulationId': data.models.simulation.simulationId
+                    },
+                    data.simulationType
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo-components.js:3062]: comma

File: sirepo/package_data/static/js/sirepo-components.js [hunk 4]
```diff
@@ -3007,6 +3007,85 @@ SIREPO.app.directive('commonFooter', function() {
...
+                $scope.resetURL();
+            };
+
+            $scope.resetURL = function() {
+                $scope.newSimURL = false;
+            };
+
+            $scope.openConvertedSimulation = function() {
+                const d = appState.models;
+                d.method = $scope.convMethod;
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo-components.js:3077]: This modifies appState.models. Is that what you intend, or do you intend just to add "method" to the message. I think the "content" of the request should by something like `{method, models}`, not `{method, ...models}`.

File: sirepo/package_data/static/js/sirepo-components.js [hunk 5]
```diff
@@ -3487,8 +3566,8 @@ SIREPO.app.directive('moderationPending', function(appState, panelState) {
     return {
         restrict: 'A',
         template: `
-          <div>Your request to access {{ appName }} has been received. For additional information, contact 
-            <a href="mailto:support@radiasoft.net">support@radiasoft.net</a>.
+          <div>Your request to access {{ appName }} has been received. For additional information, contact
+            <a href="mailto:support@sirepo.com">support@sirepo.com</a>.
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo-components.js:3570]: There's probably another place with `@radiasoft.net` so fix that too.

File: tests/template/srw_shadow_converter_test.py [hunk 6]
```diff
@@ -10,22 +10,36 @@
...
-            pkjson.dump_pretty(actual, f'{name}-shadow.json')
-            expect = _read_json_from_data_dir(f'{name}-shadow.json')
+        for name in ("crl", "gaussian", "grating"):
+            srw = _read_json_from_data_dir(f"{name}-srw.json")
+            actual = Converter().srw_to_shadow(srw.models)
+            del actual["version"]
+            actual.models.simulation.pkdel("lastModified")
+            pkjson.dump_pretty(actual, f"{name}-shadow.json")
+            expect = _read_json_from_data_dir(f"{name}-shadow.json")
             pkeq(expect, actual)
```
Comment (robnagler) [inline tests/template/srw_shadow_converter_test.py:27]: Change to use case_dirs. Just needs a different prefix

File: tests/template/srw_shadow_converter_test.py [hunk 7]
```diff
@@ -10,22 +10,36 @@
...
+def test_convert_shadow_to_srw():
+    from sirepo.template.srw_shadow_converter import Converter
+    from pykern import pkunit
+    from pykern import pkjson
+    from pykern.pkcollections import PKDict
+
+    for d in pkunit.case_dirs():
+        data = pkjson.load_any(d.join("example_shadow_data.json"))
+        actual = Converter().shadow_to_srw(data).models
+        pkjson.dump_pretty(actual, d.join("example_srw_result_data.json"))
```
Comment (robnagler) [inline tests/template/srw_shadow_converter_test.py:45]: I don't think the temporaries help anything, and simplify the naming:

```py
        pkjson.dump_pretty(
             Convert().to_srw(pkjson.load_any("in.json").models),
             "out.json",
        )
```

**Review by robnagler — CHANGES_REQUESTED** [review]

File: sirepo/template/shadow.py [hunk 8]
```diff
@@ -10,71 +10,172 @@
...
-_WIGGLER_TRAJECTORY_FILENAME = 'xshwig.sha'
+_WIGGLER_TRAJECTORY_FILENAME = "xshwig.sha"
 
 
 def stateless_compute_compute_harmonic_photon_energy(data):
     return _compute_harmonic_photon_energy(data)
 
+
+def stateful_compute_convert_to_srw(data):
+    return sirepo.template.srw_shadow_converter.Converter().shadow_to_srw(data)
```
[the remark this answers is not in the document]
Comment (gurhar1133) [reply sirepo/template/shadow.py:168]: Oops. Didn't mean to ignore, just misinterpreted what you wanted. Thanks for clarifying :)

File: sirepo/template/srw_shadow_converter.py [hunk 9]
```diff
@@ -413,8 +697,53 @@ def __multipole_to_shadow(self, srw, shadow):
...
+    def __set_source_distance(self, shadow, srw):
+        k = [
+            "coherentModesAnimation",
+            "fluxReport",
+            "intensityReport",
+            "powerDensityReport",
+            "simulation",
+            "sourceIntensityReport",
+        ]
+        for n in k:
```
Comment (mkeilman) [inline sirepo/template/srw_shadow_converter.py:745]: Still looks like `k` is used in just the one place, so:
```python
for n in (
    "coherentModesAnimation",
    ...
):
```

**Review by mkeilman — COMMENTED** [review]
One minor item

File: sirepo/package_data/static/js/sirepo-components.js (hunk 4 above)
Comment (gurhar1133) [reply sirepo/package_data/static/js/sirepo-components.js:3077]: `{method, ...models}` works but `{method, models}` doesn't.

what do you think about 
```python
requestSender.sendStatefulCompute(
        appState,
        createNewSim,
        {
                method: $scope.convMethod,
                 ...appState.models,
         }
);
```

?

**Comment by moellep** [comment]
A few things I found when testing:
- open the shadow example "Beamline Examples/Complete Beamline" and selecting "Open as a New SRW Simulation" caused a server error: AttributeError: 'PKDict' object has no attribute 'f_refract'
- shadow example "Beamline Examples/Virtual Beamline" and converting to SRW. The resulting simulation had no simulation.name value.

File: sirepo/template/shadow.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gurhar1133) [reply sirepo/template/shadow.py:168]: to clarify:
`import sirepo.template.srw_shadow`
or 
`from sirepo.template.srw_shadow import Convert`
?

File: sirepo/package_data/static/js/sirepo-components.js (hunk 4 above)
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo-components.js]: What I meant was that the API should be changed on the Python side to support this interface. It's not good to assume that there is no model named "method", implicitly.

File: sirepo/template/shadow.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/template/shadow.py:168]: No unqualified imports so `import sirepo.template.srw_shadow`.

File: sirepo/package_data/static/js/sirepo-components.js (hunk 4 above)
Comment (gurhar1133) [reply sirepo/package_data/static/js/sirepo-components.js]: Maybe should be a separate issue? I think might overcomplicate this issue. Happy to work on it though

File: tests/template/srw_shadow_test.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+from pykern import pkio
+from pykern import pkunit
+from pykern.pkdebug import pkdc, pkdp, pkdlog, pkdexc
+import pytest
+
+
+def test_conversion():
+    from sirepo.template.srw_shadow import Convert
+
+    for d in pkunit.case_dirs():
```
Comment (robnagler) [inline tests/template/srw_shadow_test.py:17]: Here's a subtlety about declarative naming. The directories are named to_shadow.in, and to_shadow.out. If the files in them were named in.json and out.json for both cases, you could write the code like this:

```py
pkjson.dump_pretty(
     getattr(Convert(), d.basename)(
         pkjson.load_any("in.json").models,
         "out.json",
     ),
)
```

I didn't want to hold back the review for this, just noting that the code would be a lot shorter.

Comment (gurhar1133) [reply tests/template/srw_shadow_test.py:17]: Good point 👍

**Comment by moellep** [comment]
Looks good - could you resolve the conflicts with master (probably from formatting) for the 2 js and 4 py files.

**Pull request merged**